python/ray/data/tests/test_backpressure_e2e.py

```diff
@@ -107,21 +108,31 @@ def consumer(batch):
 
     ds = ray.data.range(1, parallelism=1).materialize()
     ds = ds.map_batches(producer, batch_size=None, num_cpus=producer_num_cpus)
+    # Add a limit op in the middle, to test that ReservationOpResourceAllocator
+    # will account limit op's resource usage to the previous producer map op.
+    if insert_limit_op:
+        ds = ds.limit(num_blocks)
     ds = ds.map_batches(consumer, batch_size=None, num_cpus=consumer_num_cpus)
+    if insert_limit_op:
+        ds = ds.limit(num_blocks)
     return ds
 
 
 @pytest.mark.parametrize(
-    "cluster_cpus, cluster_obj_store_mem_mb",
+    "cluster_cpus, cluster_obj_store_mem_mb, insert_limit_op",
     [
-        (3, 500),  # CPU not enough
-        (4, 100),  # Object store memory not enough
-        (3, 100),  # Both not enough
+        (3, 500, False),  # CPU not enough
+        (3, 500, True),  # CPU not enough
+        (4, 100, False),  # Object store memory not enough
+        (4, 100, True),  # Object store memory not enough
+        (3, 100, False),  # Both not enough
+        (3, 100, True),  # Both not enough
     ],
 )
```

Comment:
Nit: To test all existing combinations with `insert_limit_op`, it might be cleaner to do another parametrize:

```
@pytest.mark.parametrize(
    "cluster_cpus", "cluster_obj_store_mem_mb",
    [
        (3, 500),
        (4, 100),
        (3, 100),
    ]
)
@pytest.mark.parametrize("insert_limit_op", [False, True])